### DISCUSSION TITLE
Wasp undefined reference

### DISCUSSION THREAD

**Discussion by aprilnovak** [idaholab/moose#24426] [Q&A Getting Started] [answered]
I am helping @philipJFpfahl compile Cardinal (with MOOSE's conda), but they are hitting a bunch of errors like this

```
/home/philip/mambaforge3/envs/moose/wasp/include/waspcore/Interpreter.i.h:206: undefined reference to `wasp::HITParser::HITParser(wasp::AbstractInterpreter&, std::istream&, std::shared_ptr<wasp::HITLexerImpl>)'
/home/philip/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/philip/mambaforge3/envs/moose/wasp/include/waspcore/Interpreter.i.h:213: undefined reference to `wasp::HITParser::parse()'
/home/philip/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/philip/mambaforge3/envs/moose/wasp/include/waspcore/Interpreter.i.h:206: undefined reference to `wasp::HITParser::~HITParser()'
/home/philip/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/philip/mambaforge3/envs/moose/wasp/include/waspcore/Interpreter.i.h:206: undefined reference to `wasp::HITParser::~HITParser()'
collect2: error: ld returned 1 exit status
gmake[1]: *** [Makefile:22: hit] Error 1
make: *** [/home/philip/Documents/DTU/MOOSE/cardinal/contrib/moose/framework/moose.mk:138: /home/philip/Documents/DTU/MOOSE/cardinal/contrib/moose/framework/contrib/hit/hit.so] Error 2
make: *** Waiting for unfinished jobs....
```

Does anyone have an idea for a workaround? Thanks!

**Comment by GiudGiud** [comment]
Hello

Did you `git clean -xfd` the MOOSE directory?
Basically we ve gone through 2 versions of wasp, so older objects can be linked to the older one

Guillaume

**Reply by aprilnovak** [reply]
@philipJFpfahl  what hash are you using of Cardinal? Can you show me

```
git log
```

and copy-paste the first couple lines here?

**Reply by philipJFpfahl** [reply]
Hello,

Thank you for the answers.

Here is the output

```
commit a5448feab86faf38793b6922bbebc9d996b597dd (HEAD -> devel, origin/devel, origin/HEAD)
Merge: 1d6d2507 ba5739b5
Author: April Novak <novak@berkeley.edu>
Date:   Thu May 18 14:35:55 2023 -0700

    Merge pull request #698 from aprilnovak/enum
    
    Remove nek_mesh parameter from NekRSSeparateDomainProblem

commit ba5739b54b1dbf7f759c17c0789cde7a364a0bdb
Author: aprilnovak <novak@berkeley.edu>
Date:   Thu May 18 14:19:33 2023 -0500

    Use enum for nekRS mesh and unexpose to user in NekRSSeparateDomainProblem.

commit 1d6d25075baa96b14dc8e3d3f52bff2485cb9120
Merge: adea4720 7d296388
Author: April Novak <novak@berkeley.edu>
Date:   Thu May 18 13:06:00 2023 -0700

    Merge pull request #699 from aprilnovak/removed-fixed
    
    Remove fixed_mesh from postprocessors

commit 7d296388a32f8d10cba1f515bcc2dcae9a965ca8
Author: aprilnovak <novak@berkeley.edu>
Date:   Thu May 18 14:03:01 2023 -0500

    Remove fixed_mesh from postprocessors

commit adea4720f0ab5d9acb225ecf3064e2bbbf4ce5b5
Merge: 50439311 2dd3ab5a
Author: April Novak <novak@berkeley.edu>
Date:   Thu May 18 11:11:15 2023 -0700

    Merge pull request #697 from aprilnovak/update-moose
    
    Update MOOSE

commit 2dd3ab5a08d3211e06aad530fac65b66e0eb850d
Author: April Novak <novak@berkeley.edu>
Date:   Thu May 18 09:58:09 2023 -0700

    Update potassium

commit 05d1e99282569069094fa409579fc9358b447eb3
Author: April Novak <novak@berkeley.edu>
Date:   Thu May 18 09:57:49 2023 -0700

    Update iapws

commit a2ab477d968e77956693d8c81f9a6538d2b702f5
Author: April Novak <novak@berkeley.edu>
Date:   Thu May 18 09:57:30 2023 -0700
:

```

**Reply by philipJFpfahl** [reply]
@GiudGiud 
I deleted the old MOOSE dir and reinstalled conda and then set up a new MOOSE environment. 

Thank you
Philip

**Reply by aprilnovak** [reply]
@GiudGiud, I already had Philip also do a completely fresh clone of Cardinal, so there shouldn't be any non-versioned files (e.g. `git clean`) hanging around. I also tried to reproduce his error on my machine (Mac) with this Cardinal hash, but could not.

**Reply by aprilnovak** [reply]
@philipJFpfahl, you are using the MOOSE submodule in Cardinal, right? I just want to check that you are not manually setting `MOOSE_DIR` to point somewhere else? If you are, then that version may be in need of updating.

**Reply by GiudGiud** [reply]
If it s not moose I d go wipe the conda environment in case something got messed up there.

**Reply by GiudGiud** [reply]
If it s not moose I d go wipe the conda environment in case something got messed up there.

**Reply by philipJFpfahl** [reply]
@aprilnovak can you explain what you mean by the MOOSE submodule in Cardinal? I don't point to a new dir deliberately. moose should be installed in ```../moose```
The MOOSE version in my conda environment should be new.

```
commit d38411d7b2d343afaa53afc1e36373f82aae49dc (HEAD -> master, origin/master)
Merge: 18ac9ad003 b9c8e604ca
Author: moosetest <bounces@inl.gov>
Date:   Fri May 12 23:20:44 2023 -0600

    Merge commit 'b9c8e604ca417526aacf5779b79226c
```

**Reply by aprilnovak** [reply]
@philipJFpfahl can you try

```
conda deactivate
conda env remove -n moose
conda create -n moose
<open a new terminal window>
conda activate moose
mamba install moose-libmesh moose-tools
```

and then repeat those steps (starting from deleting the Cardinal directory) I listed on the Cardinal discussion page?

**Reply by GiudGiud** [reply]
I think we can use moose-dev now instead moose-libmesh moose-tools

**Reply by philipJFpfahl** [reply]
@aprilnovak 
should I do the steps (from deleting the Cardinal directory) with  ```conda activate moose``` or  ```mamba activate moose```

**Reply by aprilnovak** [reply]
Those should be equivalent - `mamba` is just a faster version of `conda`.

**Reply by philipJFpfahl** [reply]
Okay in both cases I get the following output file:

[out.txt](https://github.com/idaholab/moose/files/11520879/out.txt)

And this output:

```
cardinal$ make -j4 > out.txt
CMake Warning (dev) at /home/philip/mambaforge3/envs/moose/share/cmake-3.25/Modules/FindPackageHandleStandardArgs.cmake:438 (message):
  The package name passed to `find_package_handle_standard_args` (PkgConfig)
  does not match the name of the calling package (LIBMESH).  This can lead to
  problems in calling code that expects `find_package` result variables
  (e.g., `_FOUND`) to follow a certain pattern.
Call Stack (most recent call first):
  /home/philip/mambaforge3/envs/moose/share/cmake-3.25/Modules/FindPkgConfig.cmake:99 (find_package_handle_standard_args)
  cmake/Modules/FindLIBMESH.cmake:17 (include)
  CMakeLists.txt:130 (find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning:
  Manually-specified variables were not used by the project:

    CMAKE_Fortran_COMPILER


make[1]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
In file included from /home/philip/Documents/DTU/MOOSE/cardinal/build/header_symlinks/OpenMCProblemBase.h:27,
                 from /home/philip/Documents/DTU/MOOSE/cardinal/build/header_symlinks/OpenMCCellAverageProblem.h:21,
                 from /home/philip/Documents/DTU/MOOSE/cardinal/build/header_symlinks/OpenMCAuxKernel.h:22,
                 from /home/philip/Documents/DTU/MOOSE/cardinal/build/header_symlinks/CellDensityAux.h:21,
                 from /home/philip/Documents/DTU/MOOSE/cardinal/src/auxkernels/CellDensityAux.C:21,
                 from /home/philip/Documents/DTU/MOOSE/cardinal/build/unity_src/auxkernels_Unity.C:3:
/home/philip/Documents/DTU/MOOSE/cardinal/install/include/openmc/tallies/filter_cell_instance.h:7:10: fatal error: gsl/gsl-lite.hpp: No such file or directory
    7 | #include <gsl/gsl-lite.hpp>
      |          ^~~~~~~~~~~~~~~~~~
compilation terminated.
make: *** [/home/philip/Documents/DTU/MOOSE/cardinal/contrib/moose/framework/build.mk:148: /home/philip/Documents/DTU/MOOSE/cardinal/build/unity_src/auxkernels_Unity.x86_64-conda-linux-gnu.opt.lo] Error 1
make: *** Waiting for unfinished jobs....
```

**Reply by aprilnovak** [reply]
Can you show me this?

```
echo $CONDA_PREFIX
echo $LD_LIBRARY_PATH
```

**Reply by aprilnovak** [reply]
I think the issue is that your build is pointing to `gsl-lite` already on your system (see these lines):

```
-- Found gsl-lite: /usr/local/lib/cmake/gsl-lite (version 0.36.1)
```

but Cardinal isn't linking against those somehow. 

Can you also please show me

```
echo $LIBRARY_PATH
```

**Reply by philipJFpfahl** [reply]
```
$ echo $LIBRARY_PATH

$ 
```
Isnt liked and I get an empty row.
While:
```
$ echo $LD_LIBRARY_PATH
/home/philip/mambaforge3/envs/moose/lib:/home/philip/mambaforge3/lib:/opt/ThirdParty-7/platforms/linux64Gcc/gperftools-svn/lib:/opt/openfoam7/platforms/linux64GccDPInt32Opt/lib/openmpi-system:/opt/ThirdParty-7/platforms/linux64GccDPInt32/lib/openmpi-system:/usr/lib/x86_64-linux-gnu/openmpi/lib:/home/philip/OpenFOAM/philip-7/platforms/linux64GccDPInt32Opt/lib:/opt/site/7/platforms/linux64GccDPInt32Opt/lib:/opt/openfoam7/platforms/linux64GccDPInt32Opt/lib:/opt/ThirdParty-7/platforms/linux64GccDPInt32/lib:/opt/openfoam7/platforms/linux64GccDPInt32Opt/lib/dummy

```

**Comment by aprilnovak** [comment]
@philipJFpfahl can you perhaps try

```
export LIBRARY_PATH=/usr/local/lib:$LIBRARY_PATH
cd cardinal
rm -rf build install
make -j8
```

@pshriwise do you know what is the correct way to find gsl?

**Reply by pshriwise** [reply]
If it's a system installation, it should be found automatically. I pulled this version of `gsl-lite` (or as close as I could get, I didn't see a `v0.36.1` here: https://github.com/gsl-lite/gsl-lite/releases) and I didn't run into any issues. 

Going into the `/build/openmc` directory and running `make clean; make VERBOSE=1 | tee omc_verbose_make.txt` might be helpful to see all of the explicit paths used to look for headers during compilation -- could be some hints there.

@philipJFpfahl once you've done that, can you up load the file here?

**Reply by philipJFpfahl** [reply]
@aprilnovak Yes,

I did:
```
rm -rf cardinal/
git clone https://github.com/neams-th-coe/cardinal.git
cd cardinal
./scripts/get-dependencies.sh
conda activate moose
export ENABLE_NEK=false
export HDF5_ROOT=$CONDA_PREFIX
export LD_LIBRARY_PATH=$CONDA_PREFIX/lib:$LD_LIBRARY_PATH
export LIBRARY_PATH=/usr/local/lib:$LIBRARY_PATH
```
with ```  echo $LIBRARY_PATH ``` I now get ``` /usr/local/lib:/usr/local/lib:```
But the error is still there.
```
fatal error: gsl/gsl-lite.hpp: No such file or directory
```

The gsl folder exists:
```
usr/local/lib/cmake/gsl-lite$ ls
gsl-lite-config.cmake  gsl-lite-config-version.cmake  gsl-lite-targets.cmake
```

**Comment by philipJFpfahl** [comment]
@pshriwise thank you for helping!

here is the output file for:

```
~/Documents/DTU/MOOSE/cardinal/build/openmc$ make clean; make VERBOSE=1 | tee omc_verbose_make.txt

```

[omc_verbose_make.txt](https://github.com/idaholab/moose/files/11523062/omc_verbose_make.txt)

**Reply by pshriwise** [reply]
Interesting! From what I can tell, it looks like that build of OpenMC was successful.

FWIW here's the set of locations in the compile line where it's looking for headers (other than the default locations the compiler will look that aren't shown in the make command):

```
/home/philip/Documents/DTU/MOOSE/cardinal/contrib/openmc/include
/home/philip/Documents/DTU/MOOSE/cardinal/build/openmc/include 
/usr/local/include
/home/philip/mambaforge3/envs/moose/libmesh/include 
/home/philip/mambaforge3/envs/moose/libmesh-vtk/include/vtk-9.2
```

**Reply by aprilnovak** [reply, marked as answer]
@philipJFpfahl, we believe this is happening due to the different build systems used by OpenMC and Cardinal. OpenMC uses CMake, whereas all MOOSE-based applications use Make. OpenMC is building fine with your system GSL, but those include dirs aren't propagated up to Cardinal, so it can't find it (this would be easy if MOOSE used CMake).

I think we have a hot fix for you, that forces OpenMC to build with its submodule GSL, which Cardinal should know how to find. Can you try editing the `cardinal/contrib/openmc/CMakeLists.txt` file to comment out this line, like:

```
#find_package_write_status(gsl-lite)
```

And then re-build:

```
cd cardinal
rm -rf build install
make -j4
```

**Reply by philipJFpfahl** [reply]
Okay, first of all, I want to say a big thank you! It is amazing that you all put so much effort into solving this. An executable has been made and it passes all the tests. There were three more libraries missing:
```
   12 | #include "pugixml.hpp"

    8 | #include "fmt/format.h"

   14 | #include "xtensor/xadapt.hpp"

```
I also found them in the CMakeLists.txt and commented them out. And now it runs like a charm.
Super cool.

**Reply by aprilnovak** [reply]
That's great! Thanks for sticking with us :) We really appreciate the feedback so we know how to make the build process more robust going forward.

**Reply by philipJFpfahl** [reply]
Much appreciate it! Without the helping hands in the forum, MOOSE would only be half as useful.
